File: app/src/main/java/ru/job4j/todolist/store/SqlStore.java
package ru.job4j.todolist.store;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import ru.job4j.todolist.Task;

public class SqlStore extends SQLiteOpenHelper implements IStore {
    public static final String DB = "todo.db";
    public static final int VERSION = 1;
    private List<Task> tasks = new ArrayList<>();
    private static SqlStore INST;

    public SqlStore(@Nullable Context context) {
        super(context, DB, null , VERSION);
        Cursor cursor = this.getWritableDatabase().query(
                TodoDbSchema.TodoTable.NAME,
                null,
                null, null,
                null, null, null
        );
        cursor.moveToFirst();
        while (!cursor.isAfterLast()) {
            this.tasks.add(new Task(
                    cursor.getInt(cursor.getColumnIndex("id")),
                    cursor.getString(cursor.getColumnIndex("name")),
                    cursor.getString(cursor.getColumnIndex("desc")),
                    cursor.getString(cursor.getColumnIndex("created")),
                    cursor.getString(cursor.getColumnIndex("closed"))
            ));
            cursor.moveToNext();
        }
        cursor.close();
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
                db.execSQL(
                "create table " +
                        TodoDbSchema.TodoTable.NAME + " (" +
                        "id integer primary key autoincrement, " +
                        TodoDbSchema.TodoTable.Cols.NAME + ", " +
                        TodoDbSchema.TodoTable.Cols.DESC + ", " +
                        TodoDbSchema.TodoTable.Cols.CREATED + ", " +
                        TodoDbSchema.TodoTable.Cols.CLOSED + " " +
                        ")"
        );
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
    }

    public static SqlStore getInstance(Context context) {
        if (INST == null){
            INST = new SqlStore(context);
        }
        return INST;
    }

    @Override
    public List<Task> getTasks() {
        return this.tasks;
    }

    @Override
    public List<Task> getFilteredTasks(String filter) {
        return tasks.stream()
                .filter(task -> task.getName().toLowerCase().contains(filter.toLowerCase()) ||
                                task.getDesc().toLowerCase().contains(filter.toLowerCase()) ||
                                task.getCreated().toLowerCase().contains(filter.toLowerCase()))
                .collect(Collectors.toList());
    }

    @Override
    public void addTask(String name,String description,String created) {
        ContentValues value = new ContentValues();
        value.put(TodoDbSchema.TodoTable.Cols.NAME, name);
        value.put(TodoDbSchema.TodoTable.Cols.DESC, description);
        value.put(TodoDbSchema.TodoTable.Cols.CREATED, created);
        int id = (int) this.getWritableDatabase().insert(TodoDbSchema.TodoTable.NAME, null, value);
        Task task = new Task(id, name, description, created, null);
        this.tasks.add(task);
    }

    @Override
    public void editTask(int id, String name, String description, String closed) {
        ContentValues value = new ContentValues();
        value.put(TodoDbSchema.TodoTable.Cols.NAME, name);
        value.put(TodoDbSchema.TodoTable.Cols.DESC, description);
        value.put(TodoDbSchema.TodoTable.Cols.CLOSED, closed);
        this.getWritableDatabase().update(TodoDbSchema.TodoTable.NAME,
                value,
                "id = ?",
                new String[]{String.valueOf(id)});
        Task task = findTaskByID(id);
        task.setName(name);
        task.setDesc(description);
        task.setClosed(closed);
    }

    @Override
    public void closeOrReopenTask(int id, String closed) {
        ContentValues value = new ContentValues();
        value.put(TodoDbSchema.TodoTable.Cols.CLOSED, closed);
        this.getWritableDatabase().update(TodoDbSchema.TodoTable.NAME,
                value,
                "id = ?",
                new String[]{String.valueOf(id)});
        Task task = findTaskByID(id);
        task.setClosed(closed);
    }

    @Override
    public int deleteTask(int id) {
        this.getWritableDatabase().delete(TodoDbSchema.TodoTable.NAME,
                "id = ?",
                new String[]{String.valueOf(id)}
                );
        int result = getPositionOfTaskById(id);
        tasks.remove(result);
        return result;
    }

    @Override
    public void deleteAllTasks() {
        this.getWritableDatabase().delete(TodoDbSchema.TodoTable.NAME, null, null);
        this.tasks.clear();
    }

    @Override
    public Task findTaskByID(int id) {
        return tasks.get(getPositionOfTaskById(id));
    }

    @Override
    public int getPositionOfTaskById(int id) {
        for (int index = 0 ; index < tasks.size(); index ++) {
            if (tasks.get(index).getId() == id) {
                return index;
            }
        }
        return -1;
    }
}

File: app/src/main/java/ru/job4j/todolist/store/SqlStore.kt
package ru.job4j.todolist.store

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import ru.job4j.todolist.Task
import java.util.*
import java.util.stream.Collectors

class SqlStore(context: Context) : SQLiteOpenHelper(context, DB, null, VERSION), IStore {
    private val tasks: MutableList<Task> = ArrayList()

    companion object {
        const val DB = "todo.db"
        const val VERSION = 1
        private lateinit var INST: SqlStore
        fun getInstance(context: Context): SqlStore {
            if (!this::INST.isInitialized) {
                INST = SqlStore(context)
                INST.updateStore()
            }
            return INST
        }
    }

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL(
                "create table " +
                        TodoDbSchema.TodoTable.NAME + " (" +
                        "id integer primary key autoincrement, " +
                        TodoDbSchema.TodoTable.Cols.NAME + ", " +
                        TodoDbSchema.TodoTable.Cols.DESC + ", " +
                        TodoDbSchema.TodoTable.Cols.CREATED + ", " +
                        TodoDbSchema.TodoTable.Cols.CLOSED + ", " +
                        TodoDbSchema.TodoTable.Cols.PHOTO + " " +
                        ")"
        )
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {}

    private fun updateStore(): Unit =
            writableDatabase.query(TodoDbSchema.TodoTable.NAME,
                    null,
                    null, null,
                    null, null, null
            ).run {
                moveToFirst()
                while (!isAfterLast) {
                    tasks.add(Task(
                            getInt(getColumnIndex("id")),
                            getString(getColumnIndex(TodoDbSchema.TodoTable.Cols.NAME)),
                            getString(getColumnIndex(TodoDbSchema.TodoTable.Cols.DESC)),
                            getString(getColumnIndex(TodoDbSchema.TodoTable.Cols.CREATED)),
                            getString(getColumnIndex(TodoDbSchema.TodoTable.Cols.CLOSED)),
                            getString(getColumnIndex(TodoDbSchema.TodoTable.Cols.PHOTO))
                    ))
                    moveToNext()
                }
                close()
            }


    override fun getTasks(): List<Task> = tasks

    override fun getFilteredTasks(filter: String): List<Task> = tasks.stream()
            .filter { (_, name, desc, created) ->
                name.toLowerCase(Locale.getDefault()).contains(filter.toLowerCase(Locale.getDefault())) ||
                        desc.toLowerCase(Locale.getDefault()).contains(filter.toLowerCase(Locale.getDefault())) ||
                        created.toLowerCase(Locale.getDefault()).contains(filter.toLowerCase(Locale.getDefault()))
            }
            .collect(Collectors.toList())


    override fun addTask(name: String, description: String, created: String, photo: String?) {
        val value = ContentValues().also {
            it.put(TodoDbSchema.TodoTable.Cols.NAME, name)
            it.put(TodoDbSchema.TodoTable.Cols.NAME, name)
            it.put(TodoDbSchema.TodoTable.Cols.DESC, description)
            it.put(TodoDbSchema.TodoTable.Cols.PHOTO, photo)
        }
        val id = writableDatabase.insert(TodoDbSchema.TodoTable.NAME, null, value).toInt()
        Task(id, name, description, created, null, photo).let { tasks.add(it) }
    }

    override fun editTask(id: Int, name: String, description: String, closed: String, photo: String) {
        val value = ContentValues().also {
            it.put(TodoDbSchema.TodoTable.Cols.NAME, name)
            it.put(TodoDbSchema.TodoTable.Cols.NAME, name)
            it.put(TodoDbSchema.TodoTable.Cols.DESC, description)
            it.put(TodoDbSchema.TodoTable.Cols.CLOSED, closed)
            it.put(TodoDbSchema.TodoTable.Cols.PHOTO, photo)
        }
        writableDatabase.update(TodoDbSchema.TodoTable.NAME, value, "id = ?", arrayOf(id.toString()))
        findTaskByID(id)?.also {
            it.name = name
            it.desc = name
            it.closed = name
            it.photo = name
        }
    }

    override fun closeOrReopenTask(id: Int, closed: String?) {
        val value = ContentValues().also {
            it.put(TodoDbSchema.TodoTable.Cols.CLOSED, closed)
        }
        writableDatabase.update(TodoDbSchema.TodoTable.NAME, value, "id = ?", arrayOf(id.toString()))
        findTaskByID(id)?.closed = closed
    }

    override fun deleteTask(id: Int): Int =
            tasks.indexOf(findTaskByID(id)).apply {
                writableDatabase.delete(TodoDbSchema.TodoTable.NAME, "id = ?", arrayOf(id.toString()))
                tasks.removeAt(this)
            }

    override fun deleteAllTasks() {
        writableDatabase.delete(TodoDbSchema.TodoTable.NAME, null, null)
        tasks.clear()
    }

    override fun findTaskByID(id: Int): Task? = tasks.firstOrNull { task -> task.id == id }
}